sdks/go/pkg/beam/core/runtime/exec/sdf_test.go

```diff
@@ -0,0 +1,408 @@
+	"github.com/apache/beam/sdks/go/pkg/beam/core/typex"
+	"github.com/google/go-cmp/cmp"
+	"testing"
+)
+
+// testTimestamp is a constant used to check that timestamps are retained.
+const testTimestamp = 15
+
+// testWindow is a constant used to check that windows are retained
+var testWindows = []typex.Window{window.IntervalWindow{Start: 10, End: 20}}
+
+// TestSdfNodes verifies that the various SDF nodes fulfill each of their
+// described contracts, that they each successfully invoke any SDF methods
+// needed, and that they preserve timestamps and windows correctly.
+func TestSdfNodes(t *testing.T) {
+	// Setup. The DoFns created below are defined in sdf_invokers_test.go and
+	// have testable behavior to confirm that they got correctly invoked.
+	// Without knowing the expected behavior of these DoFns, the desired outputs
+	// in the unit tests below will not make much sense.
+	dfn, err := graph.NewDoFn(&Sdf{}, graph.NumMainInputs(graph.MainSingle))
+	if err != nil {
+		t.Fatalf("invalid function: %v", err)
+	}
+	kvdfn, err := graph.NewDoFn(&KvSdf{}, graph.NumMainInputs(graph.MainKv))
+	if err != nil {
+		t.Fatalf("invalid function: %v", err)
+	}
+
+	// Validate PairWithRestriction matches its contract and properly invokes
+	// SDF method CreateInitialRestriction.
+	t.Run("PairWithRestriction", func(t *testing.T) {
+		tests := []struct {
+			name string
+			fn   *graph.DoFn
+			in   *FullValue
+			want *FullValue
+		}{
+			{
+				name: "SingleElem",
+				fn:   dfn,
+				in: &FullValue{
+					Elm:       5,
+					Elm2:      nil,
+					Timestamp: testTimestamp,
+					Windows:   testWindows,
+				},
+				want: &FullValue{
+					Elm: &FullValue{
+						Elm:       5,
+						Elm2:      nil,
+						Timestamp: testTimestamp,
+						Windows:   testWindows,
+					},
+					Elm2:      Restriction{5},
+					Timestamp: testTimestamp,
+					Windows:   testWindows,
+				},
+			},
+			{
+				name: "KvElem",
+				fn:   kvdfn,
+				in: &FullValue{
+					Elm:       5,
+					Elm2:      2,
+					Timestamp: testTimestamp,
+					Windows:   testWindows,
+				},
+				want: &FullValue{
+					Elm: &FullValue{
+						Elm:       5,
+						Elm2:      2,
+						Timestamp: testTimestamp,
+						Windows:   testWindows,
+					},
+					Elm2:      Restriction{7},
+					Timestamp: testTimestamp,
+					Windows:   testWindows,
+				},
+			},
+		}
+		for _, test := range tests {
+			test := test
+			t.Run(test.name, func(t *testing.T) {
+				ctx := context.Background()
+				fake := &FakeNode{}
+				node := PairWithRestriction{UID: 0, Fn: test.fn, Out: []Node{fake}}
+
+				if err := node.Up(ctx); err != nil {
+					t.Fatalf("Up failed: %v", err)
+				}
+				if err := node.StartBundle(ctx, "bundle_id", DataContext{}); err != nil {
+					t.Fatalf("StartBundle failed: %v", err)
+				}
+
+				if err := node.ProcessElement(ctx, test.in); err != nil {
+					t.Fatalf("ProcessElement failed: %v", err)
+				}
+				got := fake.Vals[0]
+				if !cmp.Equal(got, test.want) {
```

Comment:
I forgotwe had cmp probably configured in the gogradle.lock file. Nice. 

 https://github.com/apache/beam/blob/master/sdks/go/gogradle.lock#L223

In a different PR we could probably update the comparison helpers in fullvalue_test.go to use cmp options and Transformers instead which would make things much clearer.